### PR TITLE
Analysis/prompt experiments

### PR DESCRIPTION
**Pull request by curt-tigges**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,4 +11,9 @@ ccs_act_patching*.html
 *.html
 !data
 neel-plotly
-.vscode/
\ No newline at end of file
+.vscode/
+
```
Comment (ojh31) [inline .gitignore]: Can you please add .ipynb_checkpoints here and unstage the checkpoints in this PR?

File: 1-4b-continuation-attn_head_out_act_patch_results.pkl
Comment (ojh31) [inline 1-4b-continuation-attn_head_out_act_patch_results.pkl:1]: Do we want this to live on git? I wonder if we should ignore all pickle files

File: path_patching.py [hunk 2]
```diff
@@ -847,6 +851,109 @@ def act_patch(
...
+    return {
+        node_name: t.tensor(results).reshape(list(patching_nodes.shape_values[node_name].values())) if isinstance(results[0], float) else results
+        for node_name, results in results_dict.items()
+    }
+
+
+def train_probe_at_layer_pos(model, act_folder, labels, layer, pos=-1, component='hook_resid_post', with_scaler=True, max_iter=100):
+
+    # Get activations
+    with open(os.path.join(act_folder, f"{component}_pos_{pos}_activations.pkl"), "rb") as f:
```
Comment (ojh31) [inline path_patching.py]: Should we add the model to the filename to avoid data being overwritten by mistake?

File: path_patching.py [hunk 3]
```diff
@@ -847,6 +851,109 @@ def act_patch(
...
+        logits = model(orig_input)
+        return patching_metric(logits)
+    
+
+def probe_patch(
+    model: HookedTransformer,
+    orig_input: Union[str, List[str], Int[Tensor, "batch seq_len"]],
+    patching_nodes: Union[IterNode, Node, List[Node]],
+    patching_metric: Callable,
+    probe_multiplier: int = 1,
```
Comment (ojh31) [inline path_patching.py]: Unused?

File: path_patching.py [hunk 4]
```diff
@@ -847,6 +851,109 @@ def act_patch(
...
+    return pipe, score
+
+
+def _probe_patch_single(
+    model: HookedTransformer,
+    orig_input: Union[str, List[str], Int[Tensor, "batch pos"]],
+    patching_nodes: Union[Node, List[Node]],
+    patching_metric: Callable,
+    new_cache: ActivationCache,
+    probe_multiplier: int = 1,
```
Comment (ojh31) [inline path_patching.py]: Unused?

File: utils/ablation.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
+import torch
+import numpy as np
+
+def resample_cache_component(component: torch.tensor) -> torch.tensor:
+    """Resamples a batch tensor according to the index of the first dimension"""
+
+    # set seeds
+    np.random.seed(77)
```
Comment (ojh31) [inline utils/ablation.py]: Move seed to args?

File: utils/ablation.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import torch
+import numpy as np
+
+def resample_cache_component(component: torch.tensor) -> torch.tensor:
```
Comment (ojh31) [inline utils/ablation.py]: Could add shape hinting here `jaxtyping.Float[torch.Tensor, "batch ..."]`

File: utils/ablation.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    batch_size = component.shape[0]
+
+    # Resample the batch
+    indices = np.random.choice(batch_size, batch_size, replace=True)
+    component = component[indices]
+
+    return component
+
+
+def mean_over_cache_component(component: torch.tensor) -> torch.tensor:
```
Comment (ojh31) [inline utils/ablation.py]: Shape hint here?

File: utils/ablation.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    for p in pos:
+        c[:, p, :] = cache[f"blocks.{layer}.{component_type}"][:, p, :]
+    
+    return c
+
+
+def ablate_attn_head_pos_hook(c, hook, cache, ablation_func=None, component_type="hook_z", pos=-1, layer=0, head_idx=0):
+
+    if isinstance(pos, int):
```
Comment (ojh31) [inline utils/ablation.py]: type hinting here?

File: utils/attribution.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    indices = concat_lists(
+        [slice(None)] * dim,
+        [range(shape[dim]), range(shape[dim])],
+        [slice(None)] * (len(shape) - dim - 1)
+    )
+    new_tensor[indices] = tensor
+    return new_tensor
+
+
+def from_cache(
```
Comment (ojh31) [inline utils/attribution.py]: not a very descriptive name!

File: utils/prompts.py [hunk 10]
```diff
@@ -254,6 +254,52 @@ def get_prompts(
             formatter.format(ADJ1=neutral_adjectives[i], ADJ2=neutral_adjectives[i + 1], ADJ3=neutral_adjectives[i + 2], ADJ4=neutral_top_adjectives[i], VRB=neutral_verbs[i])
             for i in range(n_prompts)
         ]
+    elif prompt_type == "res_class_1":
```
Comment (ojh31) [inline utils/prompts.py]: This needs to be migrated over to the new format

File: utils/probing.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    # batch, layer, pos, dim
+    print(f"Activation shape: {batch_list.shape}")
+    # save as pkl file
+    with open(f"data/cached_activations/2_8b_mood_inference/{component}_pos_{pos}_activations.pkl", "wb") as f:
+        pickle.dump(batch_list, f)
+
+    return batch_list
+
+
+def train_probe_at_layer_pos(model, act_folder, labels, layer, pos=0, component='hook_resid_post', with_scaler=True, max_iter=100):
```
Comment (ojh31) [inline utils/probing.py]: This function is duplicated from path_patching.py

**Review by ojh31 — CHANGES_REQUESTED** [review]
I'll help with some of my suggested changes

File: path_patching.py (hunk 3 above)
Comment (ojh31) [reply path_patching.py]: the probe_multiplier argument I mean

File: utils/attribution.py (hunk 9 above)
Comment (ojh31) [reply utils/attribution.py]: also this is printing too much out too much unlabelled stuff. Add verbose argument?

Comment (ojh31) [reply utils/attribution.py]: It's hard to read too, can you split up into separate functions?

File: utils/probing.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    score = pipe.score(X_test, y_test)  # apply scaling on testing data, without leaking training data.
+    print(f"Layer {layer} position {pos} test score: {score}")
+
+    return pipe, score
+
+
+def get_probe_direction(pipeline):
+
+    # Get the standard deviation and mean from the StandardScaler
+    scaler = pipeline.named_steps['standardscaler']
```
Comment (ojh31) [inline utils/probing.py]: Will this raise an error if `with_scaler` is false?

**Review by curt-tigges — COMMENTED** [review]
Looks good!

File: prompt_ideas.md
Comment (ojh31) [inline prompt_ideas.md:1]: This file looks empty?

File: utils/probing.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    print(f"Activation shape: {batch_list.shape}")
+    # save as pkl file
+    with open(f"data/cached_activations/2_8b_mood_inference/{component}_pos_{pos}_activations.pkl", "wb") as f:
+        pickle.dump(batch_list, f)
+
+    return batch_list
+
+
+def train_probe_at_layer_pos(
+    model: HookedTransformer, act_folder: str, labels: Int[np.ndarray, "batch"], layer: int, pos: int = 0,
```
Comment (ojh31) [inline utils/probing.py:60]: `model` appears unused? Do you want to add model to the filename to avoid different models reading/writing to the same file?

**Review by ojh31 — CHANGES_REQUESTED** [review]

File: utils/probing.py [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        scaled_coef = logistic_regression_model.coef_
+        scaled_intercept = logistic_regression_model.intercept_
+
+        # Unscale the coefficients
+        probe_coef = scaled_coef / std
+
+        # Unscale the intercept
+        probe_intercept = scaled_intercept - (scaled_coef * mean / std).sum(axis=1)
+    else:
+        logistic_regression_model = pipeline.named_steps['logisticregression']
```
Comment (ojh31) [inline utils/probing.py]: Minor: three duplicated lines from above

File: utils/probing.py [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    alpha: float = 1.0,
+    device: str = "cuda",
+    verbose: bool = False,
+) -> ActivationCache:
+    """
+    Train a probe on a given part of the model, apply the direction of the probe to the cache, and return the new cache.
+    """
+    new_cache = copy.deepcopy(cache)
+    n_layers = model.cfg.n_layers
+    seq_len = cache["blocks.0.hook_mlp_out"].shape[1]
```
Comment (ojh31) [inline utils/probing.py]: Minor: I think it's slightly bad practice to assume that the cache has a certain random key. Could you get the seq len using `component` as the key?

**Review by ojh31 — COMMENTED** [review]
This is looking much cleaner now! Thank you for taking the time to make those changes. I have another couple of minor suggestions but feel free to just merge if you disagree or are time crunched.

File: utils/probing.py (hunk 15 above)
Comment (curt-tigges) [reply utils/probing.py]: That should work--I've added the component as the key

File: utils/probing.py (hunk 14 above)
Comment (curt-tigges) [reply utils/probing.py]: Restructured this to remove repetition and inefficient ordering

**Review by ojh31 — APPROVED** [review]
New utils files look very clean and useful. Great work!
